### PR TITLE
Added !wolfram command

### PR DESCRIPTION
**Pull request by mubiquity**
A single command !wolfram that does two things. Depending on the --full arg

!wolfram --full 2x+8
!wolfram 2x+8 --full
Examples:
wolframfull

!wolfram on the other hand returns just a short text response that can usually be replied to. For example:

wolfram1
wolfram2
wolfram3
wolfram4

If Wolfram cannot come up with a conversation for that question (seems to be a rare case) then a short answer is instead supplied using their short answers API. 

The app id you get from wolfram is free for less than 2000 requests a month which I doubt would ever be exceeded.

### CODE REVIEW

**Comment by mitchmcdee** [comment]
This looks lit. I'll give a code review for later :^)

File: uqcsbot/scripts/wolfram.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+from typing import Iterable, Tuple
+import requests
+import json
+
+APP_ID = "UVKV2V-5XW2TETT69"
+
+# IDEA: Show the assumptions made?
+# TODO: Better naming of commands
+@bot.on_command("wolframfull")
+async def handle_wolframfull(command: Command):
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: maybe instead of a new command, you can handle an optional variable? looks like it would remove some code duplication as well.

i.e. !wolfram something and then !wolfram something --full

if this is too annoying, consider it an optional improvement :^)

File: uqcsbot/scripts/wolfram.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    http_response = await bot.run_async(requests.get, api_url, params={'input': search_query, 'appid': APP_ID})
+
+    # Check if the response is ok
+    if http_response.status_code != requests.codes.ok:
+        bot.post_message(command.channel, "There was a problem getting the response")
+        return
+
+    # Get the result of the query and determine if wolfram succeeded in evaluating it
+    result = json.loads(http_response.content)['queryresult']
+    if not result['success'] or result["error"]:
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: why two options? shouldn't just one of them suffice?

File: uqcsbot/scripts/wolfram.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        bot.post_message(command.channel, "Please rephrase your query. Wolfram could not compute.")
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: should  you be returning here?

File: uqcsbot/scripts/wolfram.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    # Check if the response is ok. A status code of 501 signifies that no result could be found.
+    if http_response.status_code == 501:
+        return "No short answer available. Try !wolframfull"
+    elif http_response.status_code != requests.codes.ok:
+        return "There was a problem getting the response"
+
+    return http_response.content
+
+
+def get_subpods(pods: list) -> Iterable[Tuple[str, dict]]:
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: what the heck is a subpod? if its short for something, use the elaborated version in the function name. Also, add an explanation of what a subpod is in this description

Comment (mitchmcdee) [reply uqcsbot/scripts/wolfram.py]: Also, I'd move this function to top-- it's generally easier to see helper functions before getting to where they're used rather than having to dig down into the code

File: uqcsbot/scripts/wolfram.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    """
+    Yields subpods in the order they should be displayed
+
+    Yield: (pod_or_subpod_title, subpod)
+    """
+    for pod in pods:
+        for subpod in pod["subpods"]:
+            # Use the pods title if the subpod doesn't have its own title (general case)
+            title = subpod['title'] if subpod['title'] else pod['title']
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: why not `title = subpod.get('title', pod['title'])`? Will pod always have a title?

Comment (mitchmcdee) [reply uqcsbot/scripts/wolfram.py]: Are you meant to be checking if `title` exists in subpod or if its non-empty? because I think your code would KeyError if you're checking if it exists

Comment (mitchmcdee) [reply uqcsbot/scripts/wolfram.py]: If you're checking its non-empty, then yeah my suggestion won't work hehe-- but I think in that case you should make your code more verbose i.e. `if len(subpod['title']) != 0`

File: uqcsbot/scripts/wolfram.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+    bot.post_message(channel, reply, thread_ts=thread_ts)
+
+    # Update the old fallback to reflect the new state of the conversation
+    parent_attachments['fallback'] = f'WolframCanReply {reply_host} {s_output} {conversation_id}'
+
+    bot.api.chat.update(channel=channel, attachments=[parent_attachments], ts=thread_ts)
+
+
+async def short_answer(search_query: str):
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: would prefer `get_short_answer` :)

File: uqcsbot/scripts/wolfram.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    """
+    This uses wolfram's short answers api to just return a simple short plaintext response.
+
+    This is used if the conversation api fails to get a result (for instance !wolfram pineapple is not a great
+    conversation starter but may be interesting.
+    """
+    api_url = r"http://api.wolframalpha.com/v1/result?"
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: I notice you use v1 here and v2 somewhere else, any reason?

File: uqcsbot/scripts/wolfram.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    if http_response.status_code != requests.codes.ok:
+        bot.post_message(command.channel, "There was a problem getting the response")
+        return
+
+    # Get the result of the query and determine if wolfram succeeded in evaluating it
+    result = json.loads(http_response.content)['queryresult']
+    if not result['success'] or result["error"]:
+        bot.post_message(command.channel, "Please rephrase your query. Wolfram could not compute.")
+
+    subpods = get_subpods(result['pods'])
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: this needs renaming/a descriptive comment but I'll reference it in the function comments :P

Comment (mitchmcdee) [reply uqcsbot/scripts/wolfram.py]: If you don't rename it, add comments here to be explicit I guess

File: uqcsbot/scripts/wolfram.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    message = ""
+    for title, subpod in subpods:
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: I wouldn't bother extracting subpods into its own variable-- just inline the generator

File: uqcsbot/scripts/wolfram.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+        plaintext = subpod["plaintext"]
+
+        # Prefer a plain text representation to the image
+        if plaintext != "" and plaintext != "* * * * * *":
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: why are these necessary? can you link an API query example?

File: uqcsbot/scripts/wolfram.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+            message += f'{title}: {plaintext}\n'
+        else:
+            image_url = subpod['img']['src']
+            image_title = subpod['img']['title']
+            message += f'{image_title}:\n{image_url}\n' if image_title else f'{image_url}\n'
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: I'd prefer `len(image_title) != 0` over `image_title`, its not clear that you're checking non-emptiness

File: uqcsbot/scripts/wolfram.py (hunk 1 above)
Comment (mitchmcdee) [reply uqcsbot/scripts/wolfram.py]: lmk if you think they're sufficiently different responses that it warrants splitting

File: uqcsbot/scripts/wolfram.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+    This uses wolfram's conversation api to return a short response that can be replied to in a thread.
+    If the response cannot be replied to a general short answer response is displayed instead.
+
+    Example Usage:
+    !wolfram Solve Newton's Second Law for mass
+    !wolfram What is the distance from Earth to Mars?
+
+    and then start a thread to continue to conversation
+    """
+    api_url = r"http://api.wolframalpha.com/v1/conversation.jsp"
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: why `r""`?

File: uqcsbot/scripts/wolfram.py (hunk 7 above)
Comment (mitchmcdee) [reply uqcsbot/scripts/wolfram.py]: and why `r""`?

File: uqcsbot/scripts/wolfram.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+# TODO: Better naming of commands
+@bot.on_command("wolframfull")
+async def handle_wolframfull(command: Command):
+    """
+    This posts the full results from wolfram query. Images and all
+
+    Example usage:
+    !wolframfull y = 2x + c
+    """
+    api_url = r"http://api.wolframalpha.com/v2/query?&output=json"
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: why `r""`?

File: uqcsbot/scripts/wolfram.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+        bot.post_message(command.channel, "There was a problem getting the response")
+        return
+
+    wolfram_result = json.loads(http_response.content)
+    if 'error' in wolfram_result:
+        error = wolfram_result['error']
+        if error == "No result is available":
+            # If no conversational result is available just return a normal short answer
+            short_response = await short_answer(search_query)
+            bot.post_message(command.channel, short_response)
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: I think you can remove the `await` here and inline the function call

File: uqcsbot/scripts/wolfram.py [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+            return
+        else:
+            bot.post_message(command.channel, error)
+            return
+
+    # Unpack the response in order to construct the reply message
+    result = wolfram_result['result']  # This is the answer to our question
+    reply_host = wolfram_result['host']  # This is the hostname to ask the next question to
+    conversation_id = wolfram_result['conversationID']  # Used to continue the conversation
+    s_output = wolfram_result.get('s', None)  # s is only sometimes returned but is vital if it is returned
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: if its vital, explain why and what it is! :P

File: uqcsbot/scripts/wolfram.py [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+
+    # TODO: Is there a better option than storing the id in the fallback?
+    # Here we store the conversation ID in the fallback so we can get it back later.
+    # We also store an identifier string to check against later and the reply_host and s_output string
+    attachments = [{
+        'fallback': f'WolframCanReply {reply_host} {s_output} {conversation_id}',
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: is `s_output` is None, this will have a double space, is that an issue?

File: uqcsbot/scripts/wolfram.py [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+    # Unpack the response in order to construct the reply message
+    result = wolfram_result['result']  # This is the answer to our question
+    reply_host = wolfram_result['host']  # This is the hostname to ask the next question to
+    conversation_id = wolfram_result['conversationID']  # Used to continue the conversation
+    s_output = wolfram_result.get('s', None)  # s is only sometimes returned but is vital if it is returned
+
+    # TODO: Is there a better option than storing the id in the fallback?
+    # Here we store the conversation ID in the fallback so we can get it back later.
+    # We also store an identifier string to check against later and the reply_host and s_output string
+    attachments = [{
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: explain what's the attachment used for and why

File: uqcsbot/scripts/wolfram.py (hunk 16 above)
Comment (mitchmcdee) [reply uqcsbot/scripts/wolfram.py]: also I'm happy to keep it for now, but I'm sure we can do something better than stuffing it in fallback hehe :^) maybe we can build another robot brain (essentially just an in-memory key-value store we used in our jsbot)

File: uqcsbot/scripts/wolfram.py [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+        'fallback': f'WolframCanReply {reply_host} {s_output} {conversation_id}',
+        'footer': 'Further questions may be asked',
+        'text': result,
+    }]
+
+    bot.post_message(command.channel, "", attachments=attachments)
+
+
+@bot.on('message')
+async def handle_reply(evt: dict):
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: Add a function docstring here :)

File: uqcsbot/scripts/wolfram.py [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+    # If the message isn't from a thread or is from a bot ignore it (avoid those infinite loops)
+    if 'thread_ts' not in evt or (evt.get('subtype') == 'bot_message'):
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: can remove redundant parentheses

File: uqcsbot/scripts/wolfram.py [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+        return
+
+    channel = evt['channel']
+    thread_ts = evt['thread_ts']  # This refers to time the original message
+    thread_parent = bot.api.conversations.history(channel=channel, limit=1, inclusive=True, latest=thread_ts)
+
+    if not thread_parent['ok']:
+        bot.post_message(channel, 'Sorry, something went wrong with slack', thread_ts=thread_ts)
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: so what does this section do? does it post to the thread? doesn't being able to post to the thread assume that there's a thread parent, which we just learnt wasn't ok?

File: uqcsbot/scripts/wolfram.py [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+        return
+
+    parent_message = thread_parent['messages'][0]  # Limit=1 was used so the first (and only) message is what we want
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: put this comment above the line

File: uqcsbot/scripts/wolfram.py [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+    parent_attachments = parent_message['attachments'][0]  # The message is formatted with only one attachment
+    parent_fallback = parent_attachments['fallback']
+    if 'WolframCanReply' not in parent_fallback:
+        return
+
+    # Now we can grab the conversation_id from the message and get the new question (s only sometimes appears)
+    # Recall the format of the fallback "identifier hostname s_output conversationID"
+    _, reply_host, s_output, conversation_id = parent_fallback.split(' ')
+    new_question = evt['text']  # This is the value of the message that triggered this whole response
+    s_output = '' if s_output == 'None' else s_output
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: why `'None'`? was the `s_ouput` from when it was set above meant to set it to the string `None` if there wasn't an s_output? if so, I don't think you made it a string-- also, if that was your intention, I would call it 'null' to differentiate between `None`

File: uqcsbot/scripts/wolfram.py [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+
+    # Slack annoyingly formats the reply_host link so we have to extract what we want:
+    reply_host = reply_host[1:-1].split('|')[0]
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: State how Slack formats it so its clear to anyone reading this without having to test it themselves. Provide an example if need be

File: uqcsbot/scripts/wolfram.py [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+
+    # Now we can ask Wolfram for the next answer:
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: I'm seeing a lot of this boilerplate code for making a conversation api call, checking its ok and then extracting the answer-- looks like it can be abstracted into its own function and clear a bit up maybe?

File: uqcsbot/scripts/wolfram.py [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+    wolfram_answer = json.loads(http_response.content)
+    if 'error' in wolfram_answer:
+        bot.post_message(channel, wolfram_answer['error'], thread_ts=thread_ts)
+        return
+
+    # Otherwise grab the new stuff and post the reply.
+    reply = wolfram_answer['result']
+    conversation_id = wolfram_answer['conversationID']
+    reply_host = wolfram_answer['host']
+    s_output = wolfram_answer.get('s', None)
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: this section is all duplicated code-- abstract the functionality :^)

File: uqcsbot/scripts/wolfram.py [hunk 26]
```diff
@@ -0,0 +1,192 @@
+from uqcsbot import bot, Command
+from typing import Iterable, Tuple
+import requests
+import json
+
+APP_ID = "UVKV2V-5XW2TETT69"
+
+# IDEA: Show the assumptions made?
```
[the remark this answers is not in the document]
Comment (mubiquity) [reply uqcsbot/scripts/wolfram.py]: The api returns the assumptions it made during the processing. I wasn't sure whether or not to include it. Leaning towards not as the results can be long enough as is and are usually on point.

File: uqcsbot/scripts/wolfram.py [hunk 27]
```diff
@@ -0,0 +1,192 @@
+from uqcsbot import bot, Command
+from typing import Iterable, Tuple
+import requests
+import json
+
+APP_ID = "UVKV2V-5XW2TETT69"
```
[the remark this answers is not in the document]
Comment (mubiquity) [reply uqcsbot/scripts/wolfram.py]: Yeah I will but I wasn't sure whether or not you guys would want to replace it with another id

File: uqcsbot/scripts/wolfram.py (hunk 1 above)
Comment (mubiquity) [reply uqcsbot/scripts/wolfram.py]: I was on the fence about this too. I was concerned that the usual  "- - cmd" would interfere with math input but mainly that they do do pretty different things.

File: uqcsbot/scripts/wolfram.py [hunk 28]
```diff
@@ -0,0 +1,192 @@
...
+
+# IDEA: Show the assumptions made?
+# TODO: Better naming of commands
```
[the remark this answers is not in the document]
Comment (mubiquity) [reply uqcsbot/scripts/wolfram.py]: Suggestions welcome :)

File: uqcsbot/scripts/wolfram.py (hunk 13 above)
Comment (mubiquity) [reply uqcsbot/scripts/wolfram.py]: Just a url habit of mine sometimes.

File: uqcsbot/scripts/wolfram.py (hunk 2 above)
Comment (mubiquity) [reply uqcsbot/scripts/wolfram.py]: Is this about the success/error? They mean slightly different things and I think you can have one fine and the other not fine.

File: uqcsbot/scripts/wolfram.py (hunk 22 above)
Comment (mubiquity) [reply uqcsbot/scripts/wolfram.py]: I figured it would convert to a string but null is better anyway for clarity sake.

**Comment by mubiquity** [comment]
Is there a reason I can't seem to see all the comments you left aside from in my email? I only just noticed there are way more.

File: uqcsbot/scripts/wolfram.py (hunk 10 above)
Comment (mubiquity) [reply uqcsbot/scripts/wolfram.py]: Often the plaintext will just be a bunch of stars. From my testing 6 of them was the most common. But in trying to find you an example I found results that had more than that. I can't find a reference in the documentation anywhere.

File: uqcsbot/scripts/wolfram.py (hunk 15 above)
Comment (mubiquity) [reply uqcsbot/scripts/wolfram.py]: I would if I knew but the docs only say "This parameter should only be used in the rare case that your result includes the s output parameter" and "When present, this value must be propagated to the next query to ensure continuity of conversations". They don't even explain what it stands for.

File: uqcsbot/scripts/wolfram.py (hunk 20 above)
Comment (mubiquity) [reply uqcsbot/scripts/wolfram.py]: This is just a check to ensure that the thread parent was returned. I was running into a lot of auth issues in my test slack so if we see this come up that is the likely culprit.

**Comment by mitchmcdee** [comment]
I personally never use the emails for anything more than a ping + link to the github bahahhaa

**Comment by mitchmcdee** [comment]
lmk when you want another code review (not sure if you're still changing stuff)

**Comment by mubiquity** [comment]
I'll happily change more. So just whenever you have time :)

File: uqcsbot/scripts/wolfram.py (hunk 13 above)
Comment (mitchmcdee) [reply uqcsbot/scripts/wolfram.py]: I would remove the r unless its needed :^) just to be that little bit extra clear

File: uqcsbot/scripts/wolfram.py (hunk 2 above)
Comment (mitchmcdee) [reply uqcsbot/scripts/wolfram.py]: sgtm

File: uqcsbot/scripts/wolfram.py (hunk 7 above)
Comment (mitchmcdee) [reply uqcsbot/scripts/wolfram.py]: I'd upgrade to latest API if it doesn't change anything, just for consistency and longevity :^)

File: uqcsbot/scripts/wolfram.py [hunk 29]
```diff
@@ -0,0 +1,226 @@
...
+        # Update the old fallback to reflect the new state of the conversation
+        parent_attachments['fallback'] = f'WolframCanReply {reply_host} {s_output} {conversation_id}'
+
+        bot.api.chat.update(channel=channel, attachments=[parent_attachments], ts=thread_ts)
+
+
+
+
+
+
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: remove all these newlines

File: uqcsbot/scripts/wolfram.py [hunk 30]
```diff
@@ -0,0 +1,226 @@
...
+from typing import Iterable, Tuple, Optional
+from base64 import b64decode
+import requests
+import json
+
+APP_ID = b64decode('RzU0S1VBLVVHWTdHR0hWUlg=').decode('utf-8')
+
+
+
+
```
Comment (mitchmcdee) [inline uqcsbot/scripts/wolfram.py]: remove these unnecessary new lines

**Comment by mubiquity** [comment]
The conversation api doesn't have  a v2 but the short answer one does so I have changed that.

File: uqcsbot/scripts/wolfram.py [hunk 31]
```diff
@@ -0,0 +1,218 @@
+from uqcsbot import bot, Command
+from typing import Iterable, Tuple, Optional
+from base64 import b64decode
+import requests
+import json
+
+APP_ID = b64decode('RzU0S1VBLVVHWTdHR0hWUlg=').decode('utf-8')
```
Comment (CameronAavik) [inline uqcsbot/scripts/wolfram.py]: For the jsbot we have used an environment variable for this instead of hardcoding it. Would we be able to load this from an environment variable instead? (with a link on how to get an App ID)

**Comment by mitchmcdee** [comment]
In future I'd suggest splitting this PR into two halves (!wolfram and !wolframfull) though so that it can be merged quicker, good job though keep it up :^)

**Comment by mubiquity** [comment]
...I did just change it to an arg based system though hahaha

**Comment by mitchmcdee** [comment]
bahahaha no wozzas still lgtm. also, really good use of comments throughout your code 👍 makes it much easier to understand

**Comment by mitchmcdee** [comment]
is the code ready to submit? (i.e. tested and still works?)

**Comment by mubiquity** [comment]
Yep, all tested and no issues. Though it only works with the changes from #261. I completely forgot until I had to test just then and had to re-pull the changes. I think it is just some auth issue with the test slack I use which the uqcstesting slack doesn't have

**Comment by mitchmcdee** [comment]
interesting. I'll merge anyway as the code doesn't depend on #261 and the bot isn't currently in use anywho

**Comment by mitchmcdee** [comment]
oh looooooooool sorry @CameronAavik for some reason I thought this one was LGTM'd by you bahahaha
